examples/statemachine/src/language-server/statemachine-validator.ts

```diff
@@ -28,4 +33,27 @@ export class StatemachineValidator {
         }
     }
 
+    /**
+     * Checks if there are duplicate state and event names.
+     * @param statemachine the statemachine to check
+     * @param accept the acceptor to report errors
+     */
+    checkUniqueStatesAndEvents(statemachine: Statemachine, accept: ValidationAcceptor): void {
+        // check for duplicate state and event names and add them to the map
+        const names = new Map<string, State | Event>();
+        const allSymbols = [...statemachine.states, ...statemachine.events];
+        const duplicates: string[] = [];
+        for (const symbol of allSymbols) {
+            if (names.has(symbol.name)) {
+                const duplicate = names.get(symbol.name);
+                if (duplicate && !duplicates.includes(symbol.name)) {
+                    duplicates.push(symbol.name);
+                    accept('error', `Duplicate identifier name: ${symbol.name}`, { node: duplicate, property: 'name' });
+                }
+                accept('error', `Duplicate identifier name: ${symbol.name}`, { node: symbol, property: 'name' });
+            } else {
+                names.set(symbol.name, symbol);
+            }
+        }
```

Comment:
We're getting there! With `includes` we suffer again from quadratic complexity, since we need to look through the whole array of `duplicates` to confirm that an item doesn't exist in the array yet. This would be an optimized version:
```suggestion
        const names = new MultiMap<string, State | Event>();
        const allSymbols = [...statemachine.states, ...statemachine.events];
        const duplicates: string[] = [];
        for (const symbol of allSymbols) {
            names.add(symbol.name, symbol);
        }
        for (const [name, symbols] of names.entriesGroupedByKey()) {
            if (symbols.length > 1) {
                for (const symbol of symbols) {
                    accept('error', `Duplicate identifier name: ${name}`, { node: symbol, property: 'name' });
                }
            }
        }
```

Reply:
Thanks! I added this approach to arithmetics and statemachine!